File: code/greedy_algorithms/src/minimum_coins/MinimumCoins.java
// Part of Cosmos by OpenGenus Foundation
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class MinimumCoins {
    private static List<Integer> minimumCoins(Integer value, List<Integer> denominations) {
        List<Integer> result = new ArrayList<>();

        // Assuming list of denominations is sorted in descending order
        for(Integer curDenom : denominations) {
            while (curDenom <= value){
                result.add(curDenom);
                value -= curDenom;
            }
        }

        return result;
    }

    public static void main(String[] args) {
        List<TestScenario> scenarios = Arrays.asList(
            new TestScenario(100, Arrays.asList(50, 25, 10, 5, 1),      Arrays.asList(50, 50)),
            new TestScenario(101, Arrays.asList(50, 25, 10, 5, 1),      Arrays.asList(50, 50, 1)),
            new TestScenario(77,  Arrays.asList(50, 25, 10, 5, 1),      Arrays.asList(50, 25, 1, 1)),
            new TestScenario(38,  Arrays.asList(50, 25, 10, 5, 1),      Arrays.asList(25, 10, 1, 1, 1)),
            new TestScenario(17,  Arrays.asList(50, 25, 10, 5, 1),      Arrays.asList(10, 5, 1, 1)),
            new TestScenario(3,   Arrays.asList(50, 25, 10, 5, 1),      Arrays.asList(1, 1, 1)),
            new TestScenario(191, Arrays.asList(100, 50, 25, 10, 5, 1), Arrays.asList(100, 50, 25, 10, 5, 1))
        );

        for (TestScenario scenario : scenarios) {
            List<Integer> actual = minimumCoins(scenario.value, scenario.denoms);
            if (!actual.equals(scenario.result)) {
                System.out.println("Test Failed: Value: " + scenario.value
                                 + ", Denominations: " + scenario.denoms
                                 + ", Expected Result: " + scenario.result
                                 + ", Actual Result: " + actual);
            }
        }
    }

    private static class TestScenario {
        public int value;
        public List<Integer> denoms;
        public List<Integer> result;

        public TestScenario(int value, List<Integer> denoms, List<Integer> result) {
            this.value = value;
            this.denoms = denoms;
            this.result = result;
        }
    }
}

File: code/greedy_algorithms/src/minimum_coins/minimum_coins.py
# Part of Cosmos by OpenGenus Foundation


def minimum_coins(value, denominations):
    result = []
    # Assuming denominations is sorted in descendig order
    for cur_denom in denominations:
        while cur_denom <= value:
            result.append(cur_denom)
            value = value - cur_denom

    return result


# Testing
def test():
    scenarios = [
        [100, [50, 25, 10, 5, 1], [50, 50]],
        [101, [50, 25, 10, 5, 1], [50, 50, 1]],
        [77, [50, 25, 10, 5, 1], [50, 25, 1, 1]],
        [38, [50, 25, 10, 5, 1], [25, 10, 1, 1, 1]],
        [17, [50, 25, 10, 5, 1], [10, 5, 1, 1]],
        [3, [50, 25, 10, 5, 1], [1, 1, 1]],
        [191, [100, 50, 25, 10, 5, 1], [100, 50, 25, 10, 5, 1]],
    ]

    for scenario in scenarios:
        actual = minimum_coins(scenario[0], scenario[1])
        if actual != scenario[2]:
            message = "Test Failed: Value: {}, Denominations: {}, Expected Result: {}, Actual Result: {}".format(
                scenario[0], scenario[1], scenario[2], actual
            )
            print(message)


test()
